File: neet-code/math-and-geometry/spiral-matrix.py
# https://leetcode.com/problems/spiral-matrix/description/

class Solution(object):
    def spiralOrder(self, matrix):
        result = []
        rows, cols = len(matrix), len(matrix[0])
        left, right, top, bottom = 0, cols, 0, rows

        while left < right and top < bottom:
            for index in range(left, right):
                result.append(matrix[top][index])
            top += 1
            for index in range(top, bottom):
                result.append(matrix[index][right - 1])
            right -= 1

            if not(left < right and top < bottom):
                break
            
            for index in range(right - 1, left - 1, -1):
                result.append(matrix[bottom - 1][index])
            bottom -= 1
            for index in range(bottom - 1, top - 1, -1):
                result.append(matrix[index][left])
            left += 1
        
        return result

File: data-structures-1/math/spiralMatrix.js
function spiralOrder(matrix) {
  const result = [];
  const ROWS = matrix.length,
    COLS = matrix[0].length;
  let left = 0,
    right = COLS - 1,
    top = 0,
    bottom = ROWS - 1;

  while (top <= bottom && left <= right) {
    for (let i = left; i <= right; i++) {
      result.push(matrix[top][i]);
    }
    top += 1;

    for (let i = top; i <= bottom; i++) {
      result.push(matrix[i][right]);
    }
    right -= 1;

    if (!(top <= bottom && left <= right)) break;

    for (let i = right; i >= left; i--) {
      result.push(matrix[bottom][i]);
    }
    bottom -= 1;

    for (let i = bottom; i >= top; i--) {
      result.push(matrix[i][left]);
    }
    left += 1;
  }

  return result;
}
